### PR TITLE
Update iterator API

### LINKED ISSUE #1645: Simplify iterator API
Port changes from https://github.com/neo-project/neo/pull/2169, https://github.com/neo-project/neo/pull/2170, https://github.com/neo-project/neo/pull/2190 (replaces 2169).

### PR DESCRIPTION
**Pull request by fyrchik**
Close #1645 .
Remaining options TBD.

### CODE REVIEW

File: pkg/compiler/syscall_test.go [hunk 1]
```diff
@@ -4,6 +4,9 @@ import (
 	"math/big"
 	"testing"
 
+	"github.com/nspcc-dev/neo-go/pkg/core"
+	"github.com/nspcc-dev/neo-go/pkg/interop/storage"
+
```
Comment (AnnaShaleva) [inline pkg/compiler/syscall_test.go:9]: Blank line.

File: pkg/core/interop_neo.go [hunk 2]
```diff
@@ -82,11 +83,22 @@ func (s *storageIterator) Value() stackitem.Item {
 	if s.opts&FindKeysOnly != 0 {
 		return s.m[s.index].Key
 	}
+	value := s.m[s.index].Value
+	if s.opts&FindDeserialize != 0 {
+		bs, err := s.m[s.index].Value.TryBytes()
+		if err != nil {
+			panic(err) // impossible due to how map is created
+		}
```
Comment (AnnaShaleva) [inline pkg/core/interop_neo.go:99]: Then it might be just `bs := s.m[s.index].Value.Value().([]byte)`

File: pkg/interop/storage/storage.go [hunk 3]
```diff
@@ -14,6 +14,20 @@ import "github.com/nspcc-dev/neo-go/pkg/interop/iterator"
...
+
+const (
+	// None is default option.
+	None FindFlags = 0
+	// KeysOnly is used for iterating over keys.
+	KeysOnly FindFlags = 1 << 0
+	// RemovePrefix is used for stripping prefix from keys.
+	RemovePrefix FindFlags = 1 << 1
+	// ValuesOnly is used for iterating over values.
+	ValuesOnly FindFlags = 1 << 2
```
Comment (AnnaShaleva) [inline pkg/interop/storage/storage.go:28]: Can we create a test for these flags to check that they match the flags from `core` package? Like the one we had for smartcontract's triggers, it would help not to forget about them if we're to change something in `core`'s flags.

File: pkg/core/interop_neo.go [hunk 4]
```diff
@@ -21,22 +30,63 @@ func storageFind(ic *interop.Context) error {
...
 
 	filteredMap := stackitem.NewMap()
 	for k, v := range siMap {
-		filteredMap.Add(stackitem.NewByteArray(append(prefix, []byte(k)...)), stackitem.NewByteArray(v.Value))
+		var key []byte
+		if opts&FindRemovePrefix == 0 {
+			key = append(prefix, []byte(k)...)
+		} else {
+			key = []byte(k)
+		}
```
Comment (AnnaShaleva) [inline pkg/core/interop_neo.go:58]: Not exactly the way it is done in C# code. Here `prefix` is an array of bytes, and in C# it is just a single byte, see https://github.com/neo-project/neo/pull/2190/files#diff-2bbcf2b5eefab34a31011387ebf58e6a251ca395d800376cf6ef22bc5039475eR36-R37 .

**Review by AnnaShaleva — CHANGES_REQUESTED** [review]

File: pkg/interop/storage/storage.go (hunk 3 above)
Comment (fyrchik) [reply pkg/interop/storage/storage.go:28]: The test is here, in the same commit (`TestFindFlags`)

Comment (AnnaShaleva) [reply pkg/interop/storage/storage.go:28]: Somehow I missed that, sorry.

File: pkg/vm/interop.go [hunk 5]
```diff
@@ -112,45 +104,19 @@ func init() {
...
-	v.Estack().Push(&Element{
-		value: stackitem.NewInterop(interop),
-	})
-
-	return nil
-}
-
-// EnumeratorNext handles syscall System.Enumerator.Next.
-func EnumeratorNext(v *VM) error {
+// IteratorNext handles syscall System.Enumerator.Next.
```
[the remark this answers is not in the document]
Comment (fyrchik) [reply pkg/vm/interop.go:107]: Fixed.

File: pkg/vm/interop.go [hunk 6]
```diff
@@ -112,45 +104,19 @@ func init() {
...
-	arr := iop.Value().(enumerator)
+	arr := iop.Value().(iterator)
 	v.Estack().PushVal(arr.Next())
 
 	return nil
 }
 
-// EnumeratorValue handles syscall System.Enumerator.Value.
-func EnumeratorValue(v *VM) error {
+// IteratorValue handles syscall System.Enumerator.Value.
```
[the remark this answers is not in the document]
Comment (fyrchik) [reply pkg/vm/interop.go:116]: Fixed.

File: pkg/compiler/syscall_test.go (hunk 1 above)
Comment (fyrchik) [reply pkg/compiler/syscall_test.go]: Fixed.

File: pkg/core/interop_neo.go (hunk 2 above)
Comment (fyrchik) [reply pkg/core/interop_neo.go]: Fixed.

File: pkg/core/interop_neo.go (hunk 4 above)
Comment (fyrchik) [reply pkg/core/interop_neo.go]: Fixed.

**Comment by fyrchik** [comment]
RPC problem is no longer an issue, after neo-project/neo#2211 .

File: pkg/core/interop_neo.go [hunk 7]
```diff
@@ -19,6 +19,8 @@ const (
 	FindRemovePrefix = 1 << 1
 	FindValuesOnly   = 1 << 2
 	FindDeserialize  = 1 << 3
+	FindPick0        = 1 << 4
+	FindPick1        = 1 << 5
```
Comment (AnnaShaleva) [inline pkg/core/interop_neo.go:23]: Don't we also need to declare these flags as FindFlags (at `pkg/interop/storage/storage.go`)?

**Review by AnnaShaleva — CHANGES_REQUESTED** [review]

File: pkg/core/interop_neo.go (hunk 7 above)
Comment (fyrchik) [reply pkg/core/interop_neo.go:23]: Fixed.

**Review by AnnaShaleva — APPROVED** [review]

File: pkg/core/interop_neo.go [hunk 8]
```diff
@@ -90,6 +98,11 @@ func (s *storageIterator) Value() stackitem.Item {
 			panic(err)
 		}
 	}
+	if s.opts&FindPick0 != 0 {
+		value = value.Value().([]stackitem.Item)[0]
```
Comment (roman-khimov) [inline pkg/core/interop_neo.go:102]: Can we add a test for non-array and zero-length array cases? They should panic of course.

File: pkg/interop/storage/storage.go [hunk 9]
```diff
@@ -28,6 +28,10 @@ const (
 	ValuesOnly FindFlags = 1 << 2
 	// DeserializeValues is used for values deserializing.
 	DeserializeValues FindFlags = 1 << 3
+	// PickField0 is used to get first field in a serialized struct.
```
Comment (roman-khimov) [inline pkg/interop/storage/storage.go:31]: ...or array? I think both will work the same.

**Review by roman-khimov — CHANGES_REQUESTED** [review]

File: pkg/core/interop_neo.go [hunk 10]
```diff
@@ -21,6 +30,10 @@ func storageFind(ic *interop.Context) error {
 		return fmt.Errorf("%T is not a StorageContext", stcInterface)
 	}
 	prefix := ic.VM.Estack().Pop().Bytes()
+	opts := ic.VM.Estack().Pop().BigInt().Int64()
```
[the remark this answers is not in the document]
Comment (roman-khimov) [reply pkg/core/interop_neo.go:36]: See neo-project/neo#2231

File: examples/iterator/iterator.go [hunk 11]
```diff
@@ -9,13 +9,8 @@ import (
...
-	values := iterator.Values(iter)
-	keys := iterator.Keys(iter)
-
-	runtime.Notify("found storage values", values)
-	// For illustration purposes event is emitted with 'Any' type.
-	var typedKeys interface{} = keys
-	runtime.Notify("found storage keys", typedKeys)
-
+	for iterator.Next(iter) {
+		runtime.Notify("found storage value", iterator.Value(iter))
```
[the remark this answers is not in the document]
Comment (fyrchik) [reply examples/iterator/iterator.go:13]: Fixed.

File: pkg/core/interop/iterator/interop.go [hunk 12]
```diff
@@ -10,17 +10,13 @@ func Create(ic *interop.Context) error {
 	return vm.IteratorCreate(ic.VM)
 }
 
-// Key returns current iterator key.
-func Key(ic *interop.Context) error {
-	return vm.IteratorKey(ic.VM)
+// Next advances the iterator, pushes true if is it was successful
```
[the remark this answers is not in the document]
Comment (fyrchik) [reply pkg/core/interop/iterator/interop.go:13]: Fixed.

File: pkg/core/interop/iterator/interop.go [hunk 13]
```diff
@@ -10,17 +10,13 @@ func Create(ic *interop.Context) error {
...
 
-// Keys returns keys of the iterator.
-func Keys(ic *interop.Context) error {
-	return vm.IteratorKeys(ic.VM)
-}
-
-// Values returns values of the iterator.
-func Values(ic *interop.Context) error {
-	return vm.IteratorValues(ic.VM)
+// Value returns current key-value pair.
```
[the remark this answers is not in the document]
Comment (fyrchik) [reply pkg/core/interop/iterator/interop.go:19]: Documented it a bit more.

File: pkg/interop/storage/storage.go [hunk 14]
```diff
@@ -14,6 +14,20 @@ import "github.com/nspcc-dev/neo-go/pkg/interop/iterator"
 // to Neo .net framework's StorageContext class.
 type Context struct{}
 
+// FindFlags represents parameters to `Find` iterator.
+type FindFlags byte
+
+const (
+	// None is default option.
```
[the remark this answers is not in the document]
Comment (fyrchik) [reply pkg/interop/storage/storage.go]: Fixed.

File: pkg/core/interop_neo.go [hunk 15]
```diff
@@ -35,8 +48,43 @@ func storageFind(ic *interop.Context) error {
...
+	opts  int64
+	index int
+}
+
+func (s *storageIterator) Next() bool {
+	if next := s.index + 1; next < len(s.m) {
+		s.index = next
+		return true
+	}
+	return false
```
[the remark this answers is not in the document]
Comment (fyrchik) [reply pkg/core/interop_neo.go:81]: Fixed.

File: pkg/core/interop_neo.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (fyrchik) [reply pkg/core/interop_neo.go:36]: Fixed.

File: pkg/core/interop_neo.go (hunk 8 above)
Comment (fyrchik) [reply pkg/core/interop_neo.go:102]: Fixed.

File: pkg/interop/storage/storage.go (hunk 9 above)
Comment (fyrchik) [reply pkg/interop/storage/storage.go:32]: Fixed.

File: pkg/core/interop_neo.go [hunk 16]
```diff
@@ -7,12 +7,21 @@ import (
 	"sort"
 
 	"github.com/nspcc-dev/neo-go/pkg/core/interop"
-	"github.com/nspcc-dev/neo-go/pkg/vm"
 	"github.com/nspcc-dev/neo-go/pkg/vm/stackitem"
 )
 
 var errGasLimitExceeded = errors.New("gas limit exceeded")
 
+// Storage iterator options.
```
[the remark this answers is not in the document]
Comment (fyrchik) [reply pkg/core/interop_neo.go:15]: Fixed.

File: pkg/core/interop_neo.go [hunk 17]
```diff
@@ -21,6 +24,23 @@ func storageFind(ic *interop.Context) error {
 		return fmt.Errorf("%T is not a StorageContext", stcInterface)
 	}
 	prefix := ic.VM.Estack().Pop().Bytes()
+	opts := ic.VM.Estack().Pop().BigInt().Int64()
+	if opts&^storage.FindAll != 0 {
+		return fmt.Errorf("%w: unknown flag", errFindInvalidOptions)
+	}
+	if opts&storage.FindKeysOnly != 0 && opts&storage.FindDeserialize != 0 {
```
Comment (roman-khimov) [inline pkg/core/interop_neo.go:31]: pick0/pick1?

File: pkg/core/interop/storage/find.go [hunk 18]
```diff
@@ -0,0 +1,67 @@
...
+	return &Iterator{
+		m:     m.Value().([]stackitem.MapElement),
+		opts:  opts,
+		index: -1,
+	}
+}
+
+func (s *Iterator) Next() bool {
+	s.index += 1
+	return s.index < len(s.m)
```
Comment (roman-khimov) [inline pkg/core/interop/storage/find.go:35]: Overflow?

File: pkg/core/interop_neo.go (hunk 17 above)
Comment (fyrchik) [reply pkg/core/interop_neo.go:31]: It was handled implicitly, but yeah, let it be more readable. Fixed.

File: pkg/core/interop/storage/find.go (hunk 18 above)
Comment (fyrchik) [reply pkg/core/interop/storage/find.go:35]: Fixed, though highly unprobable.

**Review by roman-khimov — APPROVED** [review]

**Comment by roman-khimov** [comment]
But linter isn't happy:
```
pkg/core/interop/storage/find.go:19:6: exported type Iterator should have comment or be unexported
pkg/core/interop/storage/find.go:25:1: exported function NewIterator should have comment or be unexported
pkg/core/interop/storage/find.go:33:1: exported method Iterator.Next should have comment or be unexported
pkg/core/interop/storage/find.go:35:3: should replace s.index += 1 with s.index++
pkg/core/interop/storage/find.go:40:1: exported method Iterator.Value should have comment or be unexported
```
